src/ToData.purs

```diff
@@ -1,37 +1,111 @@
+import Data.Symbol (class IsSymbol, reflectSymbol, SProxy(..))
 import Data.Tuple.Nested (type (/\), (/\))
 import Data.UInt (UInt)
 import Helpers (uIntToBigInt)
 import Prim.TypeError (class Fail, Text)
+import Type.Proxy (Proxy(..))
 import Types.ByteArray (ByteArray)
 import Types.PlutusData (PlutusData(Constr, Integer, List, Map, Bytes))
 
+data Day = Mon | Tue | Wed | Thurs | Fri | Sat | Sun
+
+derive instance genericDay :: G.Generic Day _
+
+instance showDay :: Show Day where
+  show = genericShow
+
+instance dayToData :: ToData Day where
+  toData = genericToData
+
 class ToData (a :: Type) where
   toData :: a -> PlutusData
 
+class ConstrIndex (a :: Type) where
+  constrToIndex :: Proxy a -> Map String Int
+
+instance dayConstrIndex :: ConstrIndex Day where
+  constrToIndex _ = Map.fromFoldable [ Tuple "Mon" 0 ]
+
+-- Generic
+class ToDataArgs a where
+  toDataArgs :: a -> Array (PlutusData)
+
+instance toDataArgsNoArguments :: ToDataArgs G.NoArguments where
+  toDataArgs _ = []
+
+instance toDataArgsArgument :: ToData a => ToDataArgs (G.Argument a) where
+  toDataArgs (G.Argument x) = [ toData x ]
+
+instance toDataArgsProduct :: (ToDataArgs a, ToDataArgs b) => ToDataArgs (G.Product a b) where
+  toDataArgs (G.Product x y) = toDataArgs x <> toDataArgs y
+
+instance toDataSum' :: (ToDataArgs largs, ToDataArgs rargs) => ToData (G.Sum (G.Constructor lname largs) (G.Constructor rname rargs)) where
+  toData (G.Inl (G.Constructor args)) = Constr zero (toDataArgs args)
+  toData (G.Inr (G.Constructor args)) = Constr one (toDataArgs args)
+
+instance toDataSum'' :: (ToDataArgs args, ToData (G.Sum l r)) => ToData (G.Sum (G.Constructor name args) (G.Sum l r)) where
+  toData (G.Inl (G.Constructor args)) = Constr zero (toDataArgs args)
+  toData (G.Inr x) = toData x
+
+genericToData
+  :: forall a rep. G.Generic a rep => ToData rep => a -> PlutusData
+genericToData = toData <<< G.from
+
+class IsIndexed (a :: Type) where
+  indexFromConstructor :: forall name. IsSymbol name => IsIndexed a => Int -> SProxy name -> Proxy a -> BigInt
+
+instance sumIsIndexed :: (IsSymbol ln, IsSymbol rn) => IsIndexed (G.Sum (G.Constructor ln la) (G.Constructor rn ra)) where
+  indexFromConstructor i sname _ =
+    if reflectSymbol sname == reflectSymbol (Proxy :: Proxy ln) then fromInt i
+    else if reflectSymbol sname == reflectSymbol (Proxy :: Proxy rn) then (fromInt i + fromInt 1)
+    else BigInt.fromInt (-1)
+
+instance sumIsIndexed' :: (IsSymbol ln, IsIndexed (G.Sum l r)) => IsIndexed (G.Sum (G.Constructor ln la) (G.Sum l r)) where
+  indexFromConstructor i constrName _ =
+    if reflectSymbol constrName == reflectSymbol (Proxy :: Proxy ln) then fromInt i
+    else indexFromConstructor (i + 1) constrName (Proxy :: Proxy (G.Sum l r))
+
+genericIndexFromConstructor
+  :: forall a rep name. G.Generic a rep => IsIndexed rep => IsSymbol name => SProxy name -> Proxy a -> BigInt
+genericIndexFromConstructor sname x = indexFromConstructor 0 sname (G.from <$> x)
```

Comment:
This is here as I was hoping to get a constructor index given an (Type, Constructor) pair to fill in for the PlutusData `Constr` index field. However, that doesn't seem doable in this setup